src/gurt/misc.c

```diff
@@ -13,26 +13,482 @@
+		.mc_name	= "recov",
+	},
+	{
+		.mc_tag		= M_EC,
+		.mc_name	= "ec",
+	},
+	{
+		.mc_tag		= M_EC_AGG,
+		.mc_name	= "ec_agg",
+	},
+	{
+		.mc_tag		= M_EC_RECOV,
+		.mc_name	= "ec_recov",
+	},
+	{
+		.mc_tag		= M_POOL,
+		.mc_name	= "pool",
+	},
+	{
+		.mc_tag		= M_CONT,
+		.mc_name	= "cont",
+	},
+	{
+		.mc_tag		= M_OBJ,
+		.mc_name	= "obj",
+	},
+	{
+		.mc_tag		= M_DTX,
+		.mc_name	= "dtx",
+	},
+	{
+		.mc_tag		= M_VOS,
+		.mc_name	= "vos",
+	},
+	{
+		.mc_tag		= M_VOS_TS,
+		.mc_name	= "vos_ts",
+	},
+	{
+		.mc_tag		= M_VOS_LRU,
+		.mc_name	= "vos_lru",
+	},
+	{
+		.mc_tag		= M_VOS_DTX,
+		.mc_name	= "vos_dtx",
+	},
+	{
+		.mc_tag		= M_RDB,
+		.mc_name	= "rdb",
+	},
+	{
+		.mc_tag		= M_TAG_MAX,
+		.mc_name	= NULL,
+	},
+};
+
+/* register thread-local counter to the counter table, so we can
+ * collect the summary, see dm_tag_query().
+ */
+static void
+dm_tls_register(int tag, struct dm_tls_counter *mtc)
+{
+	struct dm_counters	*mc;
+
+	mc = &dm_counters[tag];
+	D_ASSERTF(tag == mc->mc_tag, "Mismatched tag %s: %d/%d\n",
+		  mc->mc_name, mc->mc_tag, tag);
+
+	pthread_mutex_lock(&dm_lock);
+
+	D_ASSERT(mc->mc_last < DM_TLS_MAX);
+	D_ASSERT(mc->mc_tls_cntrs[mc->mc_last] == NULL);
+
+	mc->mc_tls_cntrs[mc->mc_last] = mtc;
+	mc->mc_last++;
+	mtc->mtc_registered = true;
+
+	pthread_mutex_unlock(&dm_lock);
+}
+
+char *
+dm_mem_tag_query(int tag, int64_t *size_p, int64_t *count_p)
+{
+	struct dm_counters  *mc;
+	int64_t		     size;
+	int64_t		     count;
+	int		     i;
+
+	if (tag >= M_TAG_MAX)
+		return NULL;
+
+	/* summarize total memory consumption from thread-local counters */
+	mc = &dm_counters[tag];
+	D_ASSERT(mc->mc_tag == tag);
+
+	pthread_mutex_lock(&dm_lock);
+	for (i = size = count = 0; i < mc->mc_last; i++) {
+		D_ASSERT(mc->mc_tls_cntrs[i]);
+		size += mc->mc_tls_cntrs[i]->mtc_size;
+		count += mc->mc_tls_cntrs[i]->mtc_count;
+	}
+	pthread_mutex_unlock(&dm_lock);
+	if (size_p)
+		*size_p = size;
+	if (count_p)
+		*count_p = count;
+
+	return mc->mc_name;
+}
+
+void
+dm_mem_dump_log(void)
+{
+	char	*name;
+	int64_t	 size;
+	int64_t	 count;
+	int	 i;
+
+	D_DEBUG(DB_DM, "Memory Consumption Status:\n");
+	for (size = count = i = 0; i < M_TAG_MAX; i++) {
+		int64_t	s, c;
+
+		name = dm_mem_tag_query(i, &s, &c);
+		D_ASSERT(name);
+
+		size += s;
+		count += c;
+		D_DEBUG(DB_DM, "%-16s: size="DF_U64"(KB) count="DF_U64"\n", name, s >> 10, c);
+	}
+	D_DEBUG(DB_DM, "ALL : size="DF_U64"(KB) count="DF_U64"\n", size >> 10, count);
+}
+
+static struct dm_header *
+dm_ptr2hdr(void *ptr)
+{
+	return ptr - sizeof(struct dm_header);
+}
+
+static void *
+dm_hdr2ptr(struct dm_header *hdr)
+{
+	return (void *)&hdr->mh_payload[0];
+}
+
+static void
+dm_free(void *ptr)
+{
+	struct dm_tls_counter	*mtc;
+	struct dm_header	*hdr;
+
+	if (!ptr)
+		return;
+
+	hdr = dm_ptr2hdr(ptr);
+	/* NB: counter might belong to a different xstream, so it's not 100% safe w/o lock
+	 * but it's for debugging and hurts nothing except returning inaccurate result.
```

Comment:
right, seems can be a problem for multi-threads.
may define below mtc_size as "ATOMIC int64_t" and use atomic_fetch_sub() here for atomicity
same for mtc_count